hopsworks-common/src/main/java/io/hops/hopsworks/common/agent/AgentController.java

```diff
@@ -0,0 +1,570 @@
+    for (final SystemCommand sc : allSystemCommands) {
+      if (sc.getStatus().equals(SystemCommandFacade.STATUS.NEW)) {
+        newSystemCommands.add(sc);
+      }
+    }
+  
+    newCondaCommands.sort(ASC_COMPARATOR);
+    newSystemCommands.sort(ASC_COMPARATOR);
+    return new HeartbeatReplyDTO(newSystemCommands, newCondaCommands);
+  }
+  
+  private void updateHostMetrics(final Hosts host, final AgentHeartbeatDTO heartbeat) {
+    host.setLastHeartbeat(new Date().getTime());
+    host.setLoad1(heartbeat.load1);
+    host.setLoad5(heartbeat.load5);
+    host.setLoad15(heartbeat.load15);
+    host.setNumGpus(heartbeat.numGpus);
+    Long previousDiskUsed = host.getDiskUsed() == null ? 0L : host.getDiskUsed();
+    host.setDiskUsed(heartbeat.diskUsed);
+    host.setDiskCapacity(heartbeat.diskCapacity);
+  
+    if (((float) previousDiskUsed) / host.getDiskCapacity() < 0.8 && ((float) host.getDiskUsed()) / host.
+        getDiskCapacity() > 0.8) {
+      String subject = "alert: hard drive full on " + host.getHostname();
+      String body = host.getHostname() + " hard drive utilisation is " + host.getDiskUsageInfo();
+      emailAlert(subject, body);
+    }
+  
+    host.setMemoryUsed(heartbeat.memoryUsed);
+    host.setMemoryCapacity(heartbeat.memoryCapacity);
+    host.setPrivateIp(heartbeat.privateIp);
+    host.setCores(heartbeat.cores);
+    hostsFacade.storeHost(host);
+  }
+  
+  private void updateServices(Hosts host, AgentHeartbeatDTO heartbeat) {
+    final String hostname = heartbeat.hostId;
+    for (final AgentServiceDTO service : heartbeat.services) {
+      final String cluster = service.cluster;
+      final String name = service.service;
+      final String group = service.group;
+      HostServices hostService = null;
+      try {
+        hostService = hostServicesFacade.find(hostname, cluster, group, name);
+      } catch (Exception ex) {
+        LOG.log(Level.FINE, "Could not find service for " + hostname + "/"
+          + cluster + "/" + group + "/" + name);
+        continue;
+      }
+      if (hostService == null) {
+        hostService = new HostServices();
+        hostService.setHost(host);
+        hostService.setCluster(cluster);
+        hostService.setGroup(group);
+        hostService.setService(name);
+        hostService.setStartTime(heartbeat.agentTime);
+      }
+      
+      final Integer pid = service.pid != null ? service.pid : -1;
+      hostService.setPid(pid);
+      if (service.status != null) {
+        if ((hostService.getStatus() == null || !hostService.getStatus().equals(Status.Started))
+            && service.status.equals(Status.Started)) {
+          hostService.setStartTime(heartbeat.agentTime);
+        }
+        hostService.setStatus(service.status);
+      } else {
+        hostService.setStatus(Status.None);
+      }
+      
+      if (service.status.equals(Status.Started)) {
+        hostService.setStopTime(heartbeat.agentTime);
```

Comment:
did you mean setStartTime? or if status.stopped?

Reply:
Nope :) I just move the code from AgentResource to AgentController. But basically the idea is that agent will send a 'Started' status when a service transitioned from 'Stopped' -> 'Started', so 'heartbeat.agentTime' is an approximation of the stop time